juju/juju
# Encode controller config for json

The following is a controller config fix whereby any time.Duration would not be correctly parsed over the wire. The problem occurs when time.Duration is sent over the wire and we hit the golang bug (missing json.Marshall on the type). The time.Duration is sent over the wire, but can not be parsed back out as the time.Duration is a float64. The controller config doesn't allow coercing float64s instead we opt to encode the time duration to a string.

This seems a bit like we're papering over the cracks even more. We can't implement a json.Marshall for the controller.Config as the type is coerced into another type at the API level. I think for this reason we probably should drop time.Duration from all configurations. If you want the time.Duration we can validate it's correct when creating the config (on the way in) and then if you want the duration, we validate it on the way out.

I think if we had this type in the schema package it would solve a lot of our problems, as time.Duration would still be encoded as a string and validate can then correctly ensure they're correct.

```go
// StringifyTimeDuration returns a Checker that accepts a string value, and 
// returns a stringify time.Duration value. Empty strings are considered empty 
// time.Duration
func StringifyTimeDuration() schema.Checker {
	return stringifyTimeDurationC{}
}

type stringifyTimeDurationC struct{}

// Coerce implements Checker Coerce method.
func (c stringifyTimeDurationC) Coerce(v interface{}, path []string) (interface{}, error) {
	if v == nil {
		return nil, error_{"string or time.Duration", v, path}
	}

	var empty time.Duration
	switch reflect.TypeOf(v).Kind() {
	case reflect.TypeOf(empty).Kind():
		return v.(time.Duration).String(), nil
	case reflect.String:
		vstr := reflect.ValueOf(v).String()
		if vstr == "" {
			return empty.String(), nil
		}
		v, err := time.ParseDuration(vstr)
		if err != nil {
			return nil, parseError(path, "duration", err)
		}
		return vstr, nil
	default:
		return nil, error_{"string or time.Duration", v, path}
	}
}
```

## Checklist


## QA steps


```sh
$ juju bootstrap lxd test --build-agent
$ juju controller-config query-tracing-threshold=1ms
$ juju ssh -m controller 0
>  sudo cat /var/lib/juju/agents/machine-0/agent.conf
```

Ensure that `querytracingthreshold: 1ms` is set and it's not `1s`

## Links

**Jira card:** JUJU-[XXXX]
